src/xcoder.rs

```diff
@@ -0,0 +1,247 @@
+use std::io::Cursor;
+use std::sync::Arc;
+
+use ::image::codecs::png::{PngDecoder, PngEncoder};
+use ::image::{ColorType, DynamicImage, ImageDecoder, Rgba, RgbaImage};
+use tokio::runtime::Runtime;
+use tokio::sync::{mpsc, oneshot};
+use tokio::task::JoinHandle;
+
+use crate::chunk::{Chunk, ChunkImage};
+use crate::Error;
+
+pub struct XcoderChannels {
+   // Decoder
+   pub decoded_chunks_rx: mpsc::UnboundedReceiver<((i32, i32), RgbaImage)>,
+
+   // Encoder
+   pub encoded_chunks_rx: mpsc::UnboundedReceiver<((i32, i32), ChunkImage)>,
+}
+
+pub struct Xcoder {
+   runtime: Arc<Runtime>,
+   decoder_quitter: Option<(oneshot::Sender<()>, JoinHandle<()>)>,
+
+   // Decoder
+   chunks_to_decode_tx: mpsc::UnboundedSender<((i32, i32), Vec<u8>)>,
+
+   // Encoder
+   encoded_chunks_tx: mpsc::UnboundedSender<((i32, i32), ChunkImage)>,
+}
+
+impl Xcoder {
+   /// The maximum size threshold for a PNG to get converted to lossy WebP before network
+   /// transmission.
+   const MAX_PNG_SIZE: usize = 32 * 1024;
+   /// The quality of encoded WebP files.
+   // Note to self in the future: the libwebp quality factor ranges from 0.0 to 100.0, not
+   // from 0.0 to 1.0.
+   // 80% is a fairly sane default that preserves most of the image's quality while still retaining a
+   // good compression ratio.
+   const WEBP_QUALITY: f32 = 80.0;
+
+   pub fn new(runtime: Arc<Runtime>) -> (Self, XcoderChannels) {
+      let (chunks_to_decode_tx, chunks_to_decode_rx) = mpsc::unbounded_channel();
+      let (decoded_chunks_tx, decoded_chunks_rx) = mpsc::unbounded_channel();
+      let (encoded_chunks_tx, encoded_chunks_rx) = mpsc::unbounded_channel();
+      let (decoder_quit_tx, decoder_quit_rx) = oneshot::channel();
+      let runtime2 = Arc::clone(&runtime);
+
+      let decode_join_handle = runtime.spawn(async move {
+         Xcoder::chunk_decoding_loop(
+            runtime2,
+            chunks_to_decode_rx,
+            decoded_chunks_tx,
+            decoder_quit_rx,
+         )
+         .await;
+      });
```

Comment:
To avoid the ugly variable name `runtime2`, this can be rewritten using a block expression, like so:
```suggestion
      let decode_join_handle = runtime.spawn({
         let runtime = Arc::clone(&runtime);
         async move {
            Xcoder::chunk_decoding_loop(
               runtime,
               chunks_to_decode_rx,
               decoded_chunks_tx,
               decoder_quit_rx,
            )
            .await;
         }
      });
```